tutorial/Part X - Train on custom dataset.md

```diff
@@ -0,0 +1,178 @@
+# Part X - Train on custom dataset
+
+A lot of effort in solving machine learning problem goes into data preparation. In this tutorial, we will show how to load and preprocess/augment data from a non trivial dataset using PyTorch tools.
+
+To give an example of this process, let's use the default demo-app and modify LocalMNISTDataDataSource.kt to load custom dataset.
+
+The overall logic of file reading is: 
+1. Read the files into butter [(link)](#Reader-init-and-restart)
+2. Filter the data with demanded scope [(link)](#Custom-data-filter)
+3. Generate batches each consists of `batchSize` of samples [(link)](#Batching-and-sampling)
+
+
+Author:
+- Pengyuan Zhou - GitHub: [@pengyuan-zhou](https://github.com/pengyuan-zhou)
+
+
+```kotlin
+package org.openmined.syft.demo.federated.datasource
+
+import android.content.res.Resources
+import org.openmined.syft.demo.R
+import org.openmined.syft.demo.federated.domain.Batch
+import java.io.BufferedReader
+import java.io.FileReader;
+import java.io.File
+```
+
+
+Define class and data readers.
+```kotlin
+class LocalMNISTDataDataSource constructor(
+private val resources: Resources
+) {
+private var trainDataReader = returnDataReader()
+private var labelDataReader = returnLabelReader()
+```
+
+
+Define the paths to train file and label file.
+If the training data and labels are in the same file, then define the paths with same value.
+```kotlin
+private var trainFile = File("/path/to/train.csv")
+private var labelFile = File("/path/to/label.csv")
+```
+
+
+If the files locate in external storage, add following lines to AndroidManifest.xml
```

Comment:
This and the next column should come under a single header so that the user understands this step and the step above are replaceable